File: frontend/src/types/history.ts
import type { Protocol } from '@/types/request'

export interface HistoryRecord {
  id: string
  requestId?: string
  workspaceId: string
  protocol: Protocol
  method: string
  url: string
  requestHeaders: Record<string, string[]>
  requestBody: string
  responseStatus: number
  responseHeaders: Record<string, string[]>
  responseBody: string
  responseSize: number
  durationMs: number
  errorMessage?: string
  createdAt: string // RFC3339
}

export type StatusKind = '2xx' | '3xx' | '4xx' | '5xx' | 'error'

export interface HistoryFilter {
  requestId: string | null
  protocols: Protocol[]
  statusKinds: StatusKind[]
  urlContains: string
}

export interface ListHistoryRequest extends HistoryFilter {
  workspaceId: string
  limit: number
  offset: number
}

export interface ListHistoryResponse {
  items: HistoryRecord[]
  totalCount: number
}

File: internal/domain/entities/history.go
package entities

import (
	"time"

	"github.com/google/uuid"
)

// History is an immutable record of one request execution.
type History struct {
	ID              uuid.UUID
	RequestID       uuid.UUID
	WorkspaceID     uuid.UUID
	Protocol        Protocol
	Method          string
	URL             string
	RequestHeaders  map[string][]string
	RequestBody     string
	ResponseStatus  int
	ResponseHeaders map[string][]string
	ResponseBody    string
	ResponseSize    int64
	DurationMs      int64
	ErrorMessage    string
	// AuthQueryKeys names the query parameters auth injected into the URL, so a
	// replay draft can strip them instead of persisting a token.
	AuthQueryKeys []string
	CreatedAt     time.Time
}
